File: src/main/java/hiiragi283/materials/api/mateial/HTMaterialKey.java
package hiiragi283.materials.api.mateial;

import com.google.common.base.CaseFormat;
import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.text.TextComponentTranslation;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public final class HTMaterialKey {

    private static final Map<String, HTMaterialKey> instances = new Object2ObjectLinkedOpenHashMap<>();

    public static @NotNull HTMaterialKey getOrCreate(@NotNull final String name) {
        return instances.computeIfAbsent(name.toLowerCase(), HTMaterialKey::new);
    }

    public final @NotNull String name;
    public final @NotNull String translationKey;
    public final @NotNull String oreDictName;

    private HTMaterialKey(@NotNull String name) {
        this.name = name;
        this.translationKey = "material.ht_materials." + name;
        this.oreDictName = CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, name);
    }

    public @NotNull String getTranslatedName() {
        return I18n.format(translationKey);
    }

    public @NotNull TextComponentTranslation getTranslatedText() {
        return new TextComponentTranslation(translationKey);
    }
}

File: src/main/kotlin/io/github/hiiragi283/bwms/api/material/HTMaterialKey.kt
package io.github.hiiragi283.bwms.api.material

import io.github.hiiragi283.bwms.api.BWMsAPI
import io.github.hiiragi283.bwms.api.HTLogger
import io.github.hiiragi283.bwms.api.extension.checkNotNull
import io.github.hiiragi283.bwms.api.property.HTPropertyHolder
import net.minecraft.client.lang.text.Text
import net.minecraft.core.lang.I18n
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Supplier

class HTMaterialKey private constructor(val name: String) : Supplier<HTPropertyHolder> {
	companion object {
		private val INSTANCES: MutableMap<String, HTMaterialKey> = ConcurrentHashMap()

		@JvmStatic
		fun of(name: String): HTMaterialKey = INSTANCES.computeIfAbsent(name, ::HTMaterialKey)
	}

	// val oreDictName: String = CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, name)

	//    Supplier    //

	var validated: Boolean = false
		internal set(value) {
			if (validated) return
			HTLogger.log { it.info("Material; $name validated!") }
			field = value
		}

	fun checkValidation(): HTMaterialKey = apply {
		check(validated) { "Material; $name is not registered!" }
	}

	private var cache: HTPropertyHolder? = null

	override fun get(): HTPropertyHolder {
		checkValidation()
		if (cache == null) {
			cache = BWMsAPI.instance.materialRegistry[this]
		}
		return cache.checkNotNull { "Material; $name is not registered!" }
	}

	//    ResourceLocation    //

	// fun getId(namespace: String = HTMaterialsAPI.MOD_ID): ResourceLocation = ResourceLocation(namespace, name)

	//    Translation    //

	val translationKey: String = "ht_material.$name"

	val translatedName: String
		get() = I18n.getInstance().translateKey(translationKey)

	val translatedText: Text
		get() = Text.text().trans(translationKey)

	//    Any    //

	override fun equals(other: Any?): Boolean = (other as? HTMaterialKey)?.name == this.name

	override fun hashCode(): Int = name.hashCode()

	override fun toString(): String = name
}
